File: tools/pkg/acltester/nebraska.go
package acltester

import (
	"context"
	"encoding/xml"
	"fmt"
	"net"
	"net/http"
	"os"

	"gopkg.in/yaml.v3"
)

type NebraskaScenario struct {
	Available   bool   `yaml:"available"`
	Version     string `yaml:"version,omitempty"`
	URL         string `yaml:"url,omitempty"`
	SHA384      string `yaml:"sha384,omitempty"`
	PackageName string `yaml:"package-name,omitempty"`
}

func LoadNebraskaScenario(path string) (*NebraskaScenario, error) {
	data, err := os.ReadFile(path)
	if err != nil {
		return nil, fmt.Errorf("failed to read Nebraska scenario %s: %w", path, err)
	}
	var scenario NebraskaScenario
	if err := yaml.Unmarshal(data, &scenario); err != nil {
		return nil, fmt.Errorf("failed to parse Nebraska scenario yaml: %w", err)
	}
	return &scenario, nil
}

type NebraskaProxy struct {
	Scenario *NebraskaScenario
}

func (p *NebraskaProxy) Handler() http.Handler {
	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		if r.Method != http.MethodPost {
			http.Error(w, "method not allowed", http.StatusMethodNotAllowed)
			return
		}
		defer r.Body.Close()
		var request omahaRequest
		if err := xml.NewDecoder(r.Body).Decode(&request); err != nil {
			http.Error(w, fmt.Sprintf("failed to parse Omaha request: %v", err), http.StatusBadRequest)
			return
		}
		appID := "test"
		if len(request.Apps) > 0 && request.Apps[0].AppID != "" {
			appID = request.Apps[0].AppID
		}
		response, err := p.Scenario.BuildResponse(appID)
		if err != nil {
			http.Error(w, err.Error(), http.StatusInternalServerError)
			return
		}
		w.Header().Set("Content-Type", "application/xml")
		_, _ = w.Write(response)
	})
}

func (p *NebraskaProxy) ListenAndServe(ctx context.Context, listenAddr string) (net.Listener, error) {
	listener, err := net.Listen("tcp", listenAddr)
	if err != nil {
		return nil, fmt.Errorf("failed to listen on %s: %w", listenAddr, err)
	}
	server := &http.Server{Handler: p.Handler()}
	go func() {
		<-ctx.Done()
		_ = server.Shutdown(context.Background())
	}()
	go func() {
		_ = server.Serve(listener)
	}()
	return listener, nil
}

func (s *NebraskaScenario) BuildResponse(appID string) ([]byte, error) {
	response := omahaResponse{
		XMLName:  xml.Name{Local: "response"},
		Protocol: "3.0",
		Server:   "tester",
		Daystart: daystart{ElapsedSeconds: 0},
		Apps: []omahaApp{{
			AppID:  appID,
			Status: "ok",
		}},
	}

	if !s.Available {
		response.Apps[0].UpdateCheck = &updateCheck{Status: "noupdate"}
	} else {
		version := s.Version
		if version == "" {
			version = "1.0.0"
		}
		packageName := s.PackageName
		if packageName == "" {
			packageName = "acl.cosi"
		}
		baseURL := s.URL
		if baseURL == "" {
			baseURL = "https://example.invalid/images/"
		}
		hash := s.SHA384
		if hash == "" {
			hash = "ignored"
		}
		response.Apps[0].UpdateCheck = &updateCheck{
			Status: "ok",
			URLs:   &urls{Entries: []urlEntry{{Codebase: baseURL}}},
			Manifest: &manifest{
				Version: version,
				Packages: &packages{Entries: []packageEntry{{
					Hash:     hash,
					Name:     packageName,
					Size:     1,
					Required: true,
				}}},
			},
		}
	}

	payload, err := xml.MarshalIndent(response, "", "  ")
	if err != nil {
		return nil, err
	}
	return append([]byte(xml.Header), payload...), nil
}

type omahaRequest struct {
	Apps []struct {
		AppID string `xml:"appid,attr"`
	} `xml:"app"`
}

type omahaResponse struct {
	XMLName  xml.Name   `xml:"response"`
	Protocol string     `xml:"protocol,attr"`
	Server   string     `xml:"server,attr"`
	Daystart daystart   `xml:"daystart"`
	Apps     []omahaApp `xml:"app"`
}

type daystart struct {
	ElapsedSeconds int `xml:"elapsed_seconds,attr"`
}

type omahaApp struct {
	AppID       string       `xml:"appid,attr"`
	Status      string       `xml:"status,attr"`
	UpdateCheck *updateCheck `xml:"updatecheck,omitempty"`
}

type updateCheck struct {
	Status   string    `xml:"status,attr"`
	URLs     *urls     `xml:"urls,omitempty"`
	Manifest *manifest `xml:"manifest,omitempty"`
}

type urls struct {
	Entries []urlEntry `xml:"url"`
}

type urlEntry struct {
	Codebase string `xml:"codebase,attr"`
}

type manifest struct {
	Version  string    `xml:"version,attr"`
	Packages *packages `xml:"packages,omitempty"`
}

type packages struct {
	Entries []packageEntry `xml:"package"`
}

type packageEntry struct {
	Hash     string `xml:"hash,attr,omitempty"`
	Name     string `xml:"name,attr"`
	Size     int    `xml:"size,attr"`
	Required bool   `xml:"required,attr"`
}

File: crates/trident-acl-agent-tester/src/nebraska.rs
//! Minimal Omaha/Nebraska test double for Harpoon.

use std::{convert::Infallible, net::SocketAddr, path::Path};

use anyhow::Context;
use http_body_util::{BodyExt, Full};
use hyper::{
    body::{Bytes, Incoming},
    http::StatusCode,
    server::conn::http1,
    service::service_fn,
    Method, Request, Response,
};
use hyper_util::rt::TokioIo;
use serde::Deserialize;
use tokio::{net::TcpListener, sync::oneshot, task::JoinHandle};

#[derive(Debug, Clone, Deserialize)]
pub struct NebraskaScenario {
    #[serde(default)]
    pub available: bool,
    pub version: Option<String>,
    pub url: Option<String>,
    pub sha384: Option<String>,
}

impl NebraskaScenario {
    pub fn load(path: &Path) -> anyhow::Result<Self> {
        let contents = std::fs::read_to_string(path)
            .with_context(|| format!("failed to read Nebraska scenario {}", path.display()))?;
        serde_yaml::from_str(&contents).context("failed to parse Nebraska scenario yaml")
    }

    fn render_response(&self) -> String {
        if !self.available {
            return indoc::formatdoc! {r#"
                <?xml version="1.0" encoding="UTF-8"?>
                <response protocol="3.0" server="tester">
                  <daystart elapsed_seconds="0"/>
                  <app appid="tester" status="ok">
                    <updatecheck status="noupdate"><urls></urls></updatecheck>
                  </app>
                </response>
            "#};
        }

        let version = self.version.as_deref().unwrap_or("1.0.0");
        let url = self.url.as_deref().unwrap_or("https://example.invalid/");
        let hash = self.sha384.as_deref().unwrap_or("ignored");
        indoc::formatdoc! {r#"
            <?xml version="1.0" encoding="UTF-8"?>
            <response protocol="3.0" server="tester">
              <daystart elapsed_seconds="0"/>
              <app appid="tester" status="ok">
                <updatecheck status="ok">
                  <urls><url codebase="{url}"/></urls>
                  <manifest version="{version}">
                    <packages>
                      <package hash="{hash}" name="acl.cosi" size="1" required="true"/>
                    </packages>
                  </manifest>
                </updatecheck>
              </app>
            </response>
        "#}
    }
}

pub struct NebraskaHandle {
    pub addr: SocketAddr,
    shutdown: Option<oneshot::Sender<()>>,
    task: JoinHandle<()>,
}

impl NebraskaHandle {
    pub fn url(&self) -> String {
        format!("http://{}", self.addr)
    }

    pub async fn shutdown(mut self) {
        if let Some(tx) = self.shutdown.take() {
            let _ = tx.send(());
        }
        let _ = self.task.await;
    }
}

pub async fn spawn(
    listen: SocketAddr,
    scenario: NebraskaScenario,
) -> anyhow::Result<NebraskaHandle> {
    let listener = TcpListener::bind(listen).await?;
    let addr = listener.local_addr()?;
    let (shutdown_tx, mut shutdown_rx) = oneshot::channel();
    let xml = scenario.render_response();

    let task = tokio::spawn(async move {
        loop {
            tokio::select! {
                _ = &mut shutdown_rx => break,
                accept = listener.accept() => {
                    let Ok((stream, _)) = accept else { break; };
                    let io = TokioIo::new(stream);
                    let xml = xml.clone();
                    tokio::spawn(async move {
                        let service = service_fn(move |request| {
                            let xml = xml.clone();
                            async move { handle_request(xml, request).await }
                        });
                        let _ = http1::Builder::new().serve_connection(io, service).await;
                    });
                }
            }
        }
    });

    Ok(NebraskaHandle {
        addr,
        shutdown: Some(shutdown_tx),
        task,
    })
}

async fn handle_request(
    xml: String,
    request: Request<Incoming>,
) -> Result<Response<Full<Bytes>>, Infallible> {
    let response = match route_request(xml, request).await {
        Ok(response) => response,
        Err(err) => text_response(StatusCode::INTERNAL_SERVER_ERROR, err.to_string()),
    };
    Ok(response)
}

async fn route_request(
    xml: String,
    request: Request<Incoming>,
) -> anyhow::Result<Response<Full<Bytes>>> {
    match *request.method() {
        Method::POST => {
            let _ = request.into_body().collect().await?;
            Ok(Response::builder()
                .status(StatusCode::OK)
                .header("content-type", "application/xml")
                .body(Full::new(Bytes::from(xml)))?)
        }
        _ => Ok(text_response(
            StatusCode::METHOD_NOT_ALLOWED,
            "method not allowed",
        )),
    }
}

fn text_response(status: StatusCode, message: impl Into<String>) -> Response<Full<Bytes>> {
    Response::builder()
        .status(status)
        .header("content-type", "text/plain; charset=utf-8")
        .body(Full::new(Bytes::from(message.into())))
        .expect("valid response")
}
